daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/Parser.scala

```diff
@@ -90,6 +85,53 @@ sealed trait Parser extends Processor {
     )
   }
 
+  def PENotEnoughBits(
+    pstate: PState,
+    sfl: SchemaFileLocation,
+    dataLoc: DataLocation,
+    neededBits: Long,
+    source: InputSourceDataInputStream,
+  ): Unit = {
+    val startPos = dataLoc.bitPos1b - 1
+    val remainingLimitedBits = {
+      if (source.bitLimit0b.isEmpty) MaybeULong.Nope
+      else {
+        val lim = source.bitLimit0b.get
+        Assert.invariant(lim >= 0)
+        val nBits = lim - startPos
+        MaybeULong(nBits)
+      }
+    }
+
+    val remainingBits = {
+      if (source.hasReachedEndOfData) {
+        val bitsAvailable = {
+          val fragmentBitsRead = startPos % 8
+          source.knownBytesAvailable * 8 - fragmentBitsRead
```

Comment:
I think this is incorrect? It only uses `startPos` to know how many fragment bits to remove, but doesn't adjust the total known bits available based on `startPos`, since `knownBytesAvailable` is based on `dis.bitPos0b`.

I think we could do something like this:
```scala
val bitsAvailable = {
  val fragmentBitsReadFromSourcePos = source.bitPos0b % 8
  val bitsAvailableFromSourcePos = source.knownBytesAvailable * 8 - fragmentBitsReadFromSourcePos
  val bitsBetweenStartPosAndSourcePos = source.bitPos0b - startPos
  val bitsAvailableFromStartPos = bitsAvailableFromSourcePos + bitsBetweenStartPosandSource
  bitsAvailableFromStartPos
}
```

There might be better variable names, but the long ones do make it clear what's being calculated and why.

So this uses the previous logic to figure out how many bits are available starting from the dis.bitPos0b, then just adds the number of bits between the `DataLocation` and the `pstate currentLocation`, since the two might not be the same. 

In fact, and it might be cleaner, you could maybe have all your old logic to calculate remainingBits (which is all dis.bitPos0b based), and then at the very end do something like this to adjust the remaining bits based on the DataLocation, e.g.

```scala
val remainingBits = ... // all old logic
val bitsBetweenStartPosAndCurPos = (dis.bitPos0b - startPos)
val remainingBitsFromStartPos = remainingBits + bitsBetweenStartPosAndCurPos 
PENotEnoughBits(..., neededBits, remainingBitsFromStartPos)
```
If `dis.bitPos0b` and `sourcePos` are the same (i.e. most cases except for things that use chunkWithBits), then `bitsBetweenStartPosAndCurPos` is just 0 and doesn't have an affect. But in the case of chunkWithBits where they aren't the same, we just add the difference between where chunksWithBits started (startPos) and where it we start calculating remainingBits from (dis.bitPos0b).

Reply:
Updated